README

SWEEP, a sound wave editor, brought to you by Conrad Parker
-----------------------------------------------------------

Sweep is an audio editor and live playback tool for GNU/Linux, BSD and
compatible systems. It supports many music and voice formats including
WAV, AIFF, Ogg Vorbis, Speex and MP3, with multichannel editing and
LADSPA effects plugins. Inside lives a pesky little virtual stylus called
Scrubby who enjoys mixing around in your files.

Sweep is designed to be intuitive and to give you full control. It includes
almost everything you would expect in a sound editor, and then some:

        * precise, vinyl like scrubbing
        * looped, reverse, and pitch-controlled playback
        * playback mixing of unlimited independent tracks
	* monitoring support for use with multiple audio devices
        * looped and reverse recording
        * internationalisation
        * multichannel and 32 bit floating point file support
	* Ogg Vorbis, MP3 and Speex compressed formats
	* sample rate conversion and channel operations
        * LADSPA 1.1 effects support
        * multiple views, discontinuous selections
        * easy keybindings, mouse wheel zooming
        * unlimited undo/redo with fully revertible edit history
        * multithreaded background processing
        * shaded peak/mean waveform rendering, multiple colour schemes


Summary of library dependencies:
--------------------------------

REQUIRED:

        * GTK+ 2.0, (version 2.4.0 or higher) standard in most 
	  distributions and available from:

	    http://www.gtk.org/

        * libsndfile-1.0.x, available from:

            http://www.zip.com.au/~erikd/libsndfile/

	* Secret Rabbit Code, aka. libsamplerate, available from:

	    http://www.mega-nerd.com/SRC/
	
OPTIONAL:

	* Ogg Vorbis libraries (high quality perceptual audio format):

	    http://www.vorbis.com/

	* libmad (MPEG audio loading), available from:

	    http://www.underbit.com/products/mad/	
	
 	* Speex library (an open speech compression format):

	    http://www.speex.org/

See the file INSTALL for building instructions. Generally all that is
required, after the above libraries have been installed, is:

	./configure
	make
	make install

If you are not sure what you need, simply run ./configure and it will probe
your system and provide a reasonably useful summary.

For information on enabling and disabling optional components, run
./configure --help

Homepage
--------

Detailed information and new releases of Sweep can be found at:

        http://sweep.sourceforge.net/

Mailing Lists
-------------

There are four mailing lists related to sweep:

sweep-announce
        which contains only announcements about new releases of sweep.
This is a moderated list. 

sweep-users
	For general discussion about using sweep

sweep-devel
        For discussion about sweep development related issues. 

sweep-i18n
	For discussion of language translation issues.

Archives of each list are available on the World Wide Web via the
Sweep homepage. You can also subscribe or unsubscribe via a web
interface.

For more information about using these lists or to subscribe or
unsubscribe via an email interface, send a message with just the word
`help' as subject or in the body, to:

        sweep-announce-request@lists.sourceforge.net
  or
        sweep-users-request@lists.sourceforge.net
  or
        sweep-devel-request@lists.sourceforge.net
  or
	sweep-i18n-request@lists.sourceforge.net

as appropriate.

Internet Relay Chat
-------------------

For usage and developer discussion, join the channel #sweep on
irc.freenode.net (established 2005).

Licence
-------

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
USA.

Acknowledgements
----------------

Thanks to Pixar Animation Studios and CSIRO Australia for supporting
the development of this project.

Contact
-------

If you have any further queries about Sweep, please contact the main
author:

Conrad Parker <conrad@vergenet.net>


NEWS
Fri April 11 2008 -- Version 0.9.3 
----------------------------------------

This is a bug fix release, containing the following changes:
		
	* Fixed bounds checking of mode in speex header. relates to:
	  http://www.ocert.org/advisories/ocert-2008-2.html
	* Fixed bug preventing device dialog being reopened when
	  closed via the window manager.

All users with speex handling enabled should upgrade.


Tue Jan 30 2007 -- Version 0.9.2
--------------------------------

New in this release:

	* ALSA is now the default driver. see README.OSS for oss
	* minimum supported gtk version raised to 2.4 
	* added the "hand tool" for panning the display with inertia
	  (Radoslaw Korzeniewski)
	* can now open multiple files via the gtk file selector
	* implemented accel editing via the usual gtk method
	* added tdb lock override (Takashi Iwai)
	* remove support for libsndfile0 (Erik de Castro Lopo)
	* remove direct flac support as it's supported by libsndfile now
	  (Erik de Castro Lopo)
	* added a new splash screen (Pascal Klein)
	* fix segfault when closing windows while others were still loading
	* other bugs addressed and changes made. run "svn log" on a copy of svn
	  trunk for details, or check http://trac.metadecks.org/timeline
	* released 0.9.2


Mon Jan 30 2006 -- Version 0.9.1
--------------------------------

This is a maintenance release.

The previous release (0.9.0) had a bug that prevented sweep from
populating the ALSA devices in settings. The bug was only apparent
when you built with ALSA and didn't have a previous installation
of sweep that used ALSA.

If you installed sweep 0.9.0 and had problems getting any sound
output then please upgrade to 0.9.1.

Mon Aug 22 2005 -- Version 0.8.4
--------------------------------

This is maintenance release, including:

 	* updates for ALSA 1.0 (Daniel Dreschers, Charles Turner),
	* a new Polish translation (Radoslaw Korzeniewski),
        * an updated russian translation (Alex Prokoudine),
	* a small fix to the German translation, and
	* various minor tweaks and improvements.

This is likely to be the last Sweep release based on GTK+-1.2. Further
development will use GTK+-2.0 for the interface.

Over the last year, Peter Shorthose has developed a comprehensive set of
GTK+-2.0 patches for Sweep, and will handle their merging and ongoing
release management towards Sweep 1.0. Welcome Peter!

The trunk of sweep CVS has been migrated to a new Subversion repository.        
You can check it out by:                                                        

  svn co http://svn.metadecks.org/sweep/trunk sweep                             

svn.metadecks.org can also be browsed via trac at:                              

  http://trac.metadecks.org/                                                    


Fri Apr 30 2004 -- Version 0.8.3
--------------------------------

This is a maintenance release, including a new Spanish translation,
various bugfixes and no new functionality.

Sweep has moved back into CVS at SourceForge. Five new developers have
been added to the project: Yuri Bongiorno, Erik de Castro Lopo, Torben 
Hohn, Andr? Pang, and Alexandre Prokoudine.

Internationalization (i18n) improvements:
                                                                               
	* a new Spanish translation by Juan Manuel Martinez
        * an updated Italian translation by Yuri Bongiorno

Bugfixes:

	* resampler bug fixed by Erik de Castro Lopo 
	* made main scrollbar arrows useful (bug #821596)
	* seven other bug reports closed
	* compile warning fixes for GCC3

The dependency on libtdb has been removed. TDB sources are now pulled
directly from the upstream Samba subversion repository.


Wed Apr 30 2003 -- Version 0.8.2
--------------------------------

This release contains support for ALSA 0.9; this is configurable and could
do with widespread testing.
                                                                                
To build with ALSA support, simply configure with:
                                                                                
        ./configure --enable-alsa
                                                                                
Note that this will build a binary which will work with ALSA only, and
not attempt to use OSS. In future this will be replaced with plugins
for different pcm i/o methods, which should ease binary distribution.


Internationalization (i18n) improvements:

	* a complete audit of untranslatable strings, including
	mis-configured source files and unmarked strings, by
	Silvia Pfeiffer.

	* a new Greek translation by Dokianakis Fanis
	* a new Russian translation by Alexandre Prokoudine
	* an updated Italian translation by Yuri Bongiorno
	* an updated German translation by Silvia Pfeiffer

There is a new mailing list: sweep-i18n. List information is at
http://lists.sourceforge.net/lists/listinfo/sweep-i18n


Additional changes include:

	* added VBR without DTX option for Speex 1.0 speech encoding
	* improved accuracy of sliders in processing dialogs


Events:

Conrad will presenting and demoing Sweep at the following upcoming events:

  Jul 10-13       LinuxTag 2003 (Karlsruhe, Germany)
                  Conference and Linux Audio Dev (LAD) exhibition booth
                  http://www.linuxtag.org/
	
  Jul 31 - Aug 3  UKUUG Linux Developers Conference (Edinburgh, Scotland)
                  http://www.ukuug.org/events/linux2003/

Plus, the Call for Participation for the Linux Audio Mini-Conf @ LCA2004
(Jan 12-13 2004, Adelaide, South Australia) is announced:

    http://www.metadecks.org/events/lca2004/


Thu Feb  6 2003 -- Version 0.8.1
--------------------------------

This release contains performance improvements for basic editing operations,
including reduced memory consumption during cut and paste insert. It also
includes support for creation of new files on the command line, updated
handling of raw file loading through libsndfile, and updated support for

src/main.c
/*
 * Sweep, a sound wave editor.
 *
 * Copyright (C) 2000 Conrad Parker
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <X11/Xlib.h>

#include <glib.h>
#include <gtk/gtk.h>

#include <sweep/sweep_version.h>
#include <sweep/sweep_i18n.h>
#include <sweep/sweep_types.h>

#include "preferences.h"
#include "file_dialogs.h"
#include "interface.h"
#include "plugin.h"
#include "cursors.h"
#include "driver.h"
#include "callbacks.h"
#include "question_dialogs.h"
#include "play.h"

extern void sweep_timeouts_init (void);
extern gboolean ignore_failed_tdb_lock;
/*
 * initial_sample_load ()
 *
 * a gtk_idle one-shot function.
 *
 * the initial loading of samples is deferred
 * to be processed by gtk_main, to allow the
 * toolbox window to be shown before having to
 * wait for potentially large samples to load.
 */
static gint
initial_sample_load (gpointer data)
{
  char * arg = (char *)data;
  gchar pathname [512];

  if (!strncmp (g_dirname (arg), ".", 1)) {
    snprintf (pathname, sizeof (pathname), "%s/%s", g_get_current_dir(), arg);
  } else {
    snprintf (pathname, sizeof (pathname), "%s", arg);
  }

  sample_load (pathname);

  return FALSE;
}

static gint
initial_sample_ask (gpointer data)
{
  question_dialog_new (NULL, _("Welcome to Sweep"),
		       _("Hello, my name is Scrubby. "
		         "Welcome to Sweep!\n\n"
		         "Would you like to create a new file or "
		         "load an existing file?"),
		       _("Create new file"), _("Load existing file"),
		       G_CALLBACK (sample_new_empty_cb), NULL, G_CALLBACK (sample_load_cb), NULL, 0);

  return FALSE;
}

/*
 * main
 */
int
main (int argc, char *argv[])
{
  int i;

  gboolean show_version = FALSE;
  gboolean show_help = FALSE;
  /*  gboolean show_toolbox = TRUE;*/

  gboolean no_files = TRUE;

#ifdef HAVE_PUTENV
  gchar *display_env;
#endif

#ifdef ENABLE_NLS
  bindtextdomain (PACKAGE, PACKAGE_LOCALE_DIR);
  textdomain (PACKAGE);
#endif

  gtk_set_locale ();

#ifdef DEVEL_CODE
  g_print (_("WARNING: Build includes incomplete development code.\n"));
#endif

  XInitThreads ();

  gtk_init (&argc, &argv);

#ifdef HAVE_PUTENV
  display_env = g_strconcat ("DISPLAY=", gdk_get_display (), NULL);
  putenv (display_env);
#endif

  /* must be done before g_idle_add / g_timeout_add */
  sweep_timeouts_init ();

  for(i = 1; i < argc; i++) {
    if ((strcmp (argv[i], "--help") == 0) ||
	(strcmp (argv[i], "-h") == 0)) {
      show_help = TRUE;
      argv[i] = NULL;
    } else if ((strcmp (argv[i], "--version") == 0) ||
	       (strcmp (argv[i], "-v") == 0)) {
      show_version = TRUE;
      argv[i] = NULL;
	} else if ((strcmp (argv[i], "--ignore-failed-lock") == 0)) {
      ignore_failed_tdb_lock = TRUE;
      argv[i] = NULL;

#ifdef DEVEL_CODE
    } else if (argv[i][0] == '-' && argv[i][1] != '\0') {
      /* check for unknown options, but allow "-" as stdin */
#else
    } else if (argv[i][0] == '-') {
      /* check for unknown options */
#endif
      show_help = TRUE;
    } else {
      g_idle_add ((GSourceFunc) initial_sample_load, argv[i]);
      no_files = FALSE;
    }
  }

  if (show_version) {
    g_print ( "%s %s\n", _("Sweep version"), VERSION);
    g_print ( "%s %d.%d.%d\n",_("Sweep plugin API version"),
	      SWEEP_PLUGIN_API_MAJOR, SWEEP_PLUGIN_API_MINOR,
	      SWEEP_PLUGIN_API_REVISION);
  }

  if (show_help) {
    g_print (_("Usage: %s [option ...] [files ...]\n"), argv[0]);
    g_print (_("Valid options are:\n"));
    g_print (_("  -h --help                Output this help.\n"));
    g_print (_("  -v --version             Output version info.\n"));
    g_print (_("  --display <display>      Use the designated X display.\n"));
    g_print (_("  --ignore-failed-lock     Continue when attempt to lock the\n"
                      "                           preferences file fails.  For use when\n"
                      "                           the users home directory is on an NFS\n"
	                  "                           file system. (possibly unsafe) \n" ));

  }

  if (show_version || show_help) {
    exit (0);
  }

  srandom ((unsigned int)time(NULL));


  if (no_files) {
    g_idle_add ((GSourceFunc)initial_sample_ask, NULL);
  }

  /* initialise preferences */
  prefs_init ();

  /* initialise plugins */
  init_plugins ();

  /* initialise cursors */
  init_cursors ();

  /* initialise interface components */
  init_ui ();

  /* initialise devices */
  init_devices ();

  /* init playback subsystem */
  init_playback ();


  gtk_main ();


  /* close preferences database */
  prefs_close ();

  /* save key bindings */
  save_accels ();

  /* initialise plugins */
  release_plugins ();

  exit (0);
}
